tools/rosgraph/src/rosgraph/transport.py

```diff
@@ -0,0 +1,63 @@
+"""A replacement transport for Python xmlrpc library."""
+
+try:
+    import xmlrpc.client as xmlrpc
+except ImportError:
+    import xmlrpclib as xmlrpc
+
+import socket
+import requests
+import urllib3.exceptions
+
+
+class RequestsTransport(xmlrpc.Transport):
+    """Drop in Transport for xmlrpclib that uses Requests instead of httplib"""
+
+    user_agent = "Python XMLRPC with Requests (python-requests.org)"
+
+    def __init__(self, scheme, use_datetime=0):
+        xmlrpc.Transport.__init__(self, use_datetime)
+        self._scheme = scheme
+
+    def request(self, host, handler, request_body, verbose=0):
+        """Make a xmlrpc request."""
+        headers = {'User-Agent': self.user_agent, 'Content-Type': 'text/xml'}
+        url = '{scheme}://{host}{handler}'.format(scheme=self._scheme,
+                                                   host=host,
+                                                   handler=handler)
+        try:
+            resp = requests.post(url, data=request_body.encode('utf-8'),
+                                 headers=headers)
+        except requests.exceptions.Timeout:
+            raise socket.timeout('timed out')
+        except requests.RequestException as exc:
+            if isinstance(exc.args[0], urllib3.exceptions.HTTPError):
+                # Rethrow urllib3 exception.reason. That are e.g. socket.error
+                # Or other exceptions that the default xmlrpclib implementation
+                # would throw.
+                raise exc.args[0].reason
+            else:
+                # otherwise, rethrow the exc
+                # We could add more exception mappings here:
+                #  - requests.exceptions.InvalidURL    -> socket.gaierror?
+                #  - requests.exceptions.InvalidSchema -> socket.gaierror?
+                raise exc
```

Comment:
Since this change should address some serious problems in the current implementation I think it would be great to add. One critical question will be which ROS distro to target though. This PR is made against Indigo which I would almost rule out since it seems to be a too drastic change with too much chance for regressions for a distro already released more then two years ago.

It good target Kinetic but it would still take a lot of testing (hopefully by multiple parties) to ensure that it doesn't break anything.

The most conservative would be to only aim for the to be released L-turtle but that would mean a long time until it is actually being used (and tested in the field).

Therefore I think Kinetic is a reasonable target. We should try to maintain as much of the behavior (with the same exceptions) as reasonably possible. So if you see other extensions not being mapped yet it would be great to add them.